packages/gatsby/src/redux/nodes.js

```diff
@@ -154,3 +158,99 @@ const addResolvedNodes = (typeName, arr) => {
 }
 
 exports.addResolvedNodes = addResolvedNodes
+
+// TODO: This local cache is "global" to the nodejs instance. It should be
+//       reset after each point where nodes are mutated. This is currently
+//       not the case. I don't think we want to persist this cache in redux.
+//       (This works currently fine for a `build`, but won't for `develop`)
+let mappedByKey
+
+const ensureIndexByTypedChain = (chain, nodeTypeNames) => {
+  const chained = chain.join(`+`)
+
+  if (!mappedByKey) {
+    mappedByKey = new Map()
+  }
+
+  const nodeTypeNamePrefix = nodeTypeNames.join(`,`) + `/`
+  // The format of the typedKey is `type,type/path+to+eqobj`
+  const typedKey = nodeTypeNamePrefix + chained
+
+  let byKeyValue = mappedByKey.get(typedKey)
+  if (byKeyValue) {
+    return
+  }
+
+  console.log(`Creating new index for`, typedKey)
+  console.time(`Create cache for ` + typedKey)
+  const { nodes, resolvedNodesCache } = store.getState()
+
+  byKeyValue = new Map() // Map<node.value, Set<all nodes with this value for this key>>
+  mappedByKey.set(typedKey, byKeyValue)
+
+  // TODO: Is it faster to loop through the maps by type instead of all nodes?
+  // TODO: Are filters generally interested in internal artifact nodes that are is not related to a page?
+  nodes.forEach(node => {
+    if (!nodeTypeNames.includes(node.internal.type)) {
+      return
+    }
+
+    // There can be a filter that targets `__gatsby_resolved`
+    if (!node.__gatsby_resolved) {
+      const typeName = node.internal.type
+      const resolvedNodes = resolvedNodesCache.get(typeName)
+      node.__gatsby_resolved = resolvedNodes?.get(node.id)
+    }
+
+    // Walk the chain in the node to find the filter target
+    let v = node
+    let i = 0
+    while (i < chain.length && v) {
+      const nextProp = chain[i++]
+      v = v[nextProp]
+    }
```

Comment:
Should I do this? :)

```suggestion
    switch (chain.length) {
      case 1:
        v = node?.[chain[0]]
        break;
      case 2:
        v = node?.[chain[0]]?.[chain[1]]
        break;
      case 3:
        v = node?.[chain[0]]?.[chain[1]]?.[chain[2]]
        break;
      case 3:
        v = node?.[chain[0]]?.[chain[1]]?.[chain[3]]
        break;
      case 3:
        v = node?.[chain[0]]?.[chain[1]]?.[chain[3]]?.[chain[4]]
        break;
      default:
        // Walk the chain in the node to find the filter target
        let v = node
        let i = 0
        while (i < chain.length && v) {
          const nextProp = chain[i++]
          v = v[nextProp]
        }
    }
```